pic2img.1
\# LEGAL TERMS
\# ===========
\# This documentation is released to the public domain.
\#
\# GROFF SYNTAX
\# ============
\# Italics for files and important words, bold for programs and arguments.
\#
.
.ds timestamp 2022-04-29
.\" Update the above date whenever this file is updated.
.
.TH PIC2IMG 1 \*[timestamp] "pic2img v\*[timestamp]"
.SH NAME
pic2img \- convert a PIC diagram into a cropped image
.
.
.\" ====================================================================
.SH SYNOPSIS
.\" ====================================================================
.
.SY pic2img
.OP OPTIONS ...
.OP CONVERT-ARGUMENTS ...
.RI \< in " \>" out
.YS
.
.SY pic2img
.OP \-\-help
.OP \-\-version
.YS
.
.
.\" ====================================================================
.SH DESCRIPTION
.\" ====================================================================
.
.B pic2img
reads a
.BR gpic (1)
program from the standard input and writes an image file,
by default in Portable Network Graphics (PNG) format,
to the standard output.
.
It furthermore translates
.BR geqn (1)
constructs, so it can be used for generating images of mathematical
formulae. The 
.BR preconv (1)
preprocessor can also be enabled to parse non-English text.
.
.PP
The input PIC code should
.I not
be wrapped with the
.B \&.PS
and
.B \&.PE
macros that normally guard it within
.BR groff (1)
documents.
.
.PP
Please note that
.B pic2img
takes its input
.IR "from the standard input" ,
and 
.IR "writes to standard output" .
It does not feed from an input file or write to an output file directly.
You can use shell stream redirections (\fB\<\fR and \fB\>\fR) 
to control input and output to files.
.IP
$
.B pic2img
.RI < infile.pic " >" outfile.png
.br
.
.
.\" FIXME: How old?  This text hasn't been touched since 2008 at latest.
.\" Older versions of
.\" .I \%convert
.\" will produce a black-on-white graphic; newer ones may produce a
.\" black-on-transparent graphic.
.
.PP
Arguments not recognized by
.B pic2img
are passed to the ImageMagick program
.BR \%convert (1).
.
By specifying these, you can give your image a border,
set the image's pixel density,
or perform other useful transformations.
.
.
.PP
The output image is clipped using
.BR \%convert 's
.B \-trim
option to the smallest possible bounding box.
.
.
.\" ====================================================================
.SH OPTIONS
.\" ====================================================================
.
.
.TP
.BI \-f " format"
.TQ
.BI \-\-format " format"
Write the image in
.IR format ,
which must be understood by
.BR \%convert ;
the default is PNG.
.
.
.TP
.BI "\-q " quality
.TQ
.BI "\-\-quality " quality
Sets the output image quality, on a 1\(en3 scale (1 being the lowest quality).
The
.IR 1 ", " 2 " and " 3
values correspond to 
.BR \%convert \'s
.B -density
values of
.IR 300 ", " 600 " and " 900 
respectively, which applies to reading the input PostScript from
.BR groff .
If out of this scale, the
.I quality
value is directly mapped to the
.B -density
option.
.
.
.TP
.BI \-s " size"
Set size of the output image, according to 
.BR convert (1)'s
.B -geometry
syntax.
.
.
.TP
.B \-b
Set a white background to the output image, even when outputting to PNG 
.IR format .
.
.
.TP
.B \-k
Enable 
.BR groff 's
.BR preconv
preprocessor to parse the input.
.
.
.TP
.BI \-K " preconv-encoding"
Set
.BR groff 's
.BR preconv
preprocessor
input encoding. Implies
.BR \-k .
.
.
.TP
.BI \-\-eqn " delimiters"
Use
.I delimiters
as the opening and closing
characters that delimit
.B geqn
directives;
the default is \(lq$$\(rq.
The option argument
.I delimiters
should be a two-character string,
but an empty string (\(dq\(dq) is accepted as a directive to disable
.B geqn
processing.
.
.
.TP
.B \-U
.TQ
.B \-\-unsafe
Run
.B groff
in unsafe mode, enabling the PIC command
.B sh
to execute arbitrary Unix shell commands.
The
.BR groff (1)
default is to forbid this.
.
.
.TP
.B \-v
.TQ
.B \-\-version
Display version information and exit.
.
.
.TP
.B \-h
.TQ
.B \-\-help
Display a usage message and exit.
.
.
.\" ====================================================================
.SH "EXAMPLES"
.\" ====================================================================
.
This command creates a crisp
.I pic.jpg
JPG file of 1920 pixel of width reading the pic instructions of the
.I diagram.pic
file, with
.BR groff 's
unsafe mode and the
.B preconv
preprocessor enabled.
.IP
.RB "$ " pic2img
-b -k -U -f jpg -s 1920x -q 3
.RI < diagram.pic " >" pic.jpg
.
.
.\" ====================================================================
.SH AUTHORS
.\" ====================================================================
.
.I pic2img
was written by
.MT mail@\:sylsau.com
Sylvain Saubier
.ME ,
based on a recipe by Eric S. \& Raymonds.
.
.
.\" ====================================================================
.SH "SEE ALSO"
.\" ====================================================================
.
.PP
W.\& Richard Stevens,
.UR http://\:www.kohala.com/\:start/\:troff/\:pic2html.html
.I Turning PIC into HTML
.UE
.
.PP
ImageMagick's online manual,
.UR https://imagemagick.org/script/command-line-processing.php#geometry
.I Anatomy of the Command-line / Image Geometry
.UE
.
.PP
ImageMagick's online manual,
.UR https://imagemagick.org/script/command-line-options.php#density
.I Annotated List of Command-line Options / -density
.UE
.
.PP
.BR groff (1),
.BR \%convert (1),
.BR gpic (1),
.BR geqn (1),
.BR preconv (1),
.BR eqn2graph (1),
.BR grap2graph (1)
.
.\" Local Variables:
.\" mode: nroff
.\" End:
.\" vim: set filetype=groff:


pic2img.sh
#! /bin/sh
#
# pic2img -- compile PIC image descriptions to bitmap images
#
# by Sylvain Saubier <mail@sylsau.com>, March 2022
# based on pic2graph by Eric S. Raymond <esr@thyrsus.com>, July 2002
#
# In Unixland, the magic is in knowing what to string together...
#
# Take a pic/eqn diagram on stdin, emit cropped bitmap on stdout.
# The pic markup should *not* be wrapped in .PS/.PE, this script will do that.
# An -unsafe option on the command line enables gpic/groff "unsafe" mode.
# A --format FOO option changes the image output format to any format
# supported by convert(1).  An --eqn option changes the eqn delimiters.
# All other options are passed to convert(1).  The default format in PNG.
#
# Requires the groff suite and the ImageMagick tools.  Both are open source.
# This code is released to the public domain.
#
# Here are the assumptions behind the option processing:
#
# 1. Only a few options are relevant to groff(1)/gpic(1). -C doesn't matter
# 	because we're generating our own .PS/.PE, -[ntcz] are irrelevant because
# 	we're generating Postscript. -k/-K can be needed to handle non-english
# 	input text.
#
# 2. Many options of convert(1) are potentially relevant (especially 
# 	-background, -interlace, -border, and -comment).
#
# We don't have complete option coverage on eqn because this is primarily
# intended as a pic translator; we can live with eqn defaults. 
# We're assuming portrait letter page size is large enough to render the whole
# diagram.
# 
# The -density ImageMagick option needs to be treated separately as it
# is needed for reading the input PS file.
# 
#set -o xtrace # For debugging
set -o errexit

groff_opts=""
convert_density="-density 300"
convert_opts=""
convert_trim="-trim"
convert_size="+repage"
format="png"
eqndelim='$$'

while [ "$1" ] ; do
	case $1 in
	-U | --unsafe)
		groff_opts="$groff_opts -U";;
	-f | --format)
		format=$2
		shift;;
	-q | --quality)
		if [ $2 -eq 1 ]; then
			convert_density="-density 300"
		elif [ $2 -eq 2 ]; then
			convert_density="-density 600"
		elif [ $2 -eq 3 ]; then
			convert_density="-density 900"
		else
			convert_density="-density $2"
		fi
		shift;;
	-b)
		convert_opts="$convert_opts -background white -flatten";;
	-s)
		convert_size="-resize $2 +repage"
		shift;;
	-k)
		groff_opts="$groff_opts -k";;
	-K)
		groff_opts="$groff_opts -K $2"
		shift;;
	--eqn)
		eqndelim="$2"
		shift;;
	-v | --version)
		#echo "GNU pic2graph (groff) version @VERSION@"
		echo "pic2graph modernized edition by Sylvain Saubier, version 2022.04.29"
		exit 0;;
	-h | --help)
		echo -e "pic2img, modernized version of groff's pic2graph utility to \
turn PIC diagram into images easily.\n\
Usage:  pic2img [-f FORMAT -q QUALITY -b -s SIZE -k -K PRECONV-ENCODING --eqn EQN-DELIM -U -v] < IN > OUT\n"
		exit 0;;
	*)
		convert_opts="$convert_opts $1";;
	esac
	shift
done

if [ "$eqndelim" ] ; then
	eqndelim="delim $eqndelim"
fi

# See if installed version of convert(1) support the -trim option.
# Versions that didn't were described as "old" as early as 2008.
is_convert_recent=`convert -help | grep -e -trim`
if [ -z "$is_convert_recent" ] ; then
    echo "$0: warning: falling back to old '-crop 0x0' trim method" >&2
    convert_trim="-crop 0x0"
fi

trap 'exit_status=$? ; echo $0: error > /dev/stderr && exit $exit_status' EXIT INT TERM

# Here goes:
# 1. Wrap stdin in dummy .PS/PE macros (and add possibly null .EQ/.EN)
# 2. Process through eqn and pic to emit troff markup.
# 3. Process through groff to emit Postscript.
# 4. Use convert(1) to crop the PostScript and turn it into a bitmap.
(echo ".EQ"; echo $eqndelim; echo ".EN"; echo ".PS"; cat; echo ".PE") | \
    groff -e -p $groff_opts -Tps -P-pletter | \
    convert $convert_density - $convert_opts $convert_trim $convert_size $format:-

# End
